(DOCSP-32028): Switch to a PLP and provide info about using mongosh with Atlas (#299)
mongodb/docs-mongodb-shell: source/index.txt

:template: product-landing
:hidefeedback: header
:noprevnext:

.. program:: mongosh

.. binary:: mongosh

.. _mdb-shell-overview:

======================================
Welcome to MongoDB Shell (``mongosh``)
======================================

.. introduction::

>    The |mdb-shell|, ``mongosh``, is a JavaScript and Node.js 16.x |repl| 

Reviewer: Alternatively, I think it’s okay to leave out the version number here. Node.js major versions don’t differ as significantly as they did in the past, and for example for typical users the difference between 16 and 20 will be negligible.

   environment for interacting with MongoDB deployments in Atlas, locally,
   or on another remote host. Use the |mdb-shell| to test queries and 
   operations directly with your database.

   .. button:: Download mongosh
      :uri: https://www.mongodb.com/try/download/shell?jmp=docs

> .. image:: /images/mongosh-plp/Technical_SOFTWARE_Terminal(1)_Spot.png

Author: This is good context! Yeah, that makes a lot of sense given that feedback. I've made this change here and on the tab that was using this image.

   :alt: Abstract illustration showing a terminal interacting with different systems 

.. kicker:: What You Can Do

Access MongoDB From Your Shell
------------------------------

.. tabs::

   .. tab:: Connect to MongoDB
      :tabid: connect

>       .. procedure::

Author: I agree that the numbered procedure is a little odd here given that a user only wants to do one of these, but the styling with this product landing page template is a bit constrained. IIRC the only way to include text and image side-by-side are by using a procedure like this, or by using a `.. container::` reStructuredText directive, which technically works - we do it in the [Realm SDK docs](https://www.mongodb.com/docs/realm/sdk/swift/) on the SwiftUI/App Services tabs - but the directive is deprecated and causes build warnings.

If I switch to an Atlas-focused procedure instead of just mentioning Atlas as the first connection "option," we lose the prominent info on the landing page that you can _also_ use `mongosh` to connect to local deployments and other remote hosts. I can mention it in the text, but that would likely be missed by people who are just scanning headings to see what you can do.

With that context in mind, would you both still prefer I switch this to a procedure focused on Atlas?

(And, alas, @GaurabAryal , the GIF is an existing asset that I was able to get from the Atlas documentation set. I don't have the tools to modify it. If you think it's unclear, would you rather I remove it from this documentation set? I could put a static illustration there, instead.)

>       .. procedure::

Author: I've rewritten this tab per your suggestions, @sarahsimpers  - is this what you had in mind?

>       .. procedure::

Reviewer 2: This is exactly what I had in mind, thank you!

            including how to get the Atlas connection string, refer to 
            `Connect to a Database Deployment <https://www.mongodb.com/docs/atlas/connect-to-database-deployment/>`__
            in the Atlas documentation.

            For details about connecting to a non-default local port, 
            connecting with authentication, or other options, refer to 
            :ref:`mdb-shell-connect`.

      .. image:: /images/mongosh-plp/atlas_plp.gif
         :alt: MongoDB Atlas Connect to your Database image

   .. tab:: Perform CRUD and Aggregation
      :tabid: crud

      .. procedure::

         .. step:: Insert Documents

            :binary:`~bin.mongosh` supports common insert opererations, 
            including:

            - :method:`db.collection.insertOne()`
            - :method:`db.collection.insertMany()`

            For more information and examples, refer to :ref:`mongosh-insert`.

         .. step:: Read Documents

            Use the :method:`db.collection.find()` method to query documents
            in a collection. For more information and examples, refer to 
            :ref:`mongosh-read`.

         .. step:: Update Documents

>             :binary:`~bin.mongosh` supports common update opererations, 

Reviewer: ```suggestion
            :binary:`~bin.mongosh` supports common update operations, 
```

            including:

            - :method:`db.collection.updateOne()`
            - :method:`db.collection.updateMany()`
            - :method:`db.collection.replaceOne()`

            For more information and examples, refer to :ref:`mongosh-update`.

         .. step:: Delete Documents

>             :binary:`~bin.mongosh` supports common delete opererations, 

Reviewer: ```suggestion
            :binary:`~bin.mongosh` supports common delete operations, 
```

Author: 😅  Thank you for spotting these!

            - :method:`db.collection.deleteOne()`

            For more information and examples, refer to :ref:`mongosh-delete`.

         .. step:: Run Aggregation Pipelines

            You can run aggregation pipelines in :binary:`~bin.mongosh` 
            using the :method:`db.collection.aggregate()` method. For more 
            information and examples, refer to :ref:`mdb-shell-aggregation`.

      .. image:: /images/mongosh-plp/Technical_ACTION_SearchDocument_Thumbnail_BS_Mist.png
         :alt: Abstract illustration showing CRUD and aggregation operations

   .. tab:: Administer and Manage
      :tabid: administer

      .. procedure::

         .. step:: Manage Databases and Collections

            View information about databases, create collections or views,
            or drop databases - all from your shell. See all 
            :ref:`Database Methods <mongosh-database-methods>`.

            Perform collection operations, create or delete indexes, or 
            explain queries with :ref:`Collection Methods 
            <mongosh-collection-methods>`.

         .. step:: Administer Servers

            Manage :ref:`replication <mongosh-replication-methods>` or 

>             :ref:`sharding <mongosh-sharding-methods>` from the comfort 
>             of your shell.

Reviewer 2: Suggestion: this sounded a little odd to me, what do you think of this language instead?
```suggestion
            :ref:`sharding <mongosh-sharding-methods>` conveniently in
            your shell.
```

            Check server status with a variety of :ref:`Server Status Methods 
            <mongosh-server-status-methods>`.

         .. step:: Manage Users and Roles

            Create or update roles, define and update privileges, or drop 
            roles using :ref:`Role Management Methods 
            <mongosh-role-management-methods>`.

            Create and update users, authenticate users, and manage user
            roles with :ref:`User Management Methods 
            <mongosh-user-management-methods>`. 

      .. image:: /images/mongosh-plp/Technical_SOFTWARE_Terminal(3)_Spot.png
         :alt: Abstract illustration showing a terminal performing operations

   .. tab:: Write Scripts
      :tabid: scripts

      .. procedure::

>          .. step:: Perform Common Functions with Scripts

Reviewer 3: Can we consider breaking this section into the following parts?

1) Running DB scripts against mongosh for CRUD or administrative tasks e.g if you have a JS file that seeds synthetic / mock data into your DB in your dev env or staging, you can run the js file with `mongosh <jsfile>.js` 

2) Create custom helper functions with .mongoshrc (see [this](https://preview-mongodbdacharyc.gatsbyjs.io/mongodb-shell/DOCSP-32028/write-scripts/#execute-code-from-a-configuration-file)). The use case we should describe here could be something like: "if you convert date string to ISO format often for queries, you can create a function in `.mongoshrc` `function toISO(dateString) { return new Date(dateString).toISOString(); }` and call e.g `toISO("06/07/2017") in mongosh."

3) Use or publish snippets. You can [pull existing snippets](https://preview-mongodbdacharyc.gatsbyjs.io/mongodb-shell/DOCSP-32028/snippets/working-with-snippets/#install-snippet-packages). You can also create & publish new snippets for your custom use case, example: a snippet that validates data that you are importing daily as a cron, you can publish this as a snippet so your dev team can have access to it. You can publish either to the community registry or configure a private registry.

4) Code scoping and other considerations

Author: I've rewritten the content on this tab similar to your suggestions here, @GaurabAryal - is this what you had in mind?

            Write scripts for the |mdb-shell| that modify data in MongoDB or
            perform administrative operations.

            Explore a tutorial that uses the |mdb-shell| with JavaScript to
            access MongoDB: :ref:`mdb-shell-write-scripts`.

>          .. step:: Package and Distribute Snippets

Reviewer 3: See step 1 for change recommendation

            You may want to package and distribute your scripts as snippets
            for you and others to use in their projects. 

            Learn more about requiring external files and modules in your 
            scripts: :ref:`mongosh-require-external-files-and-modules`. 

>          .. step:: Code Scoping and Other Considerations

Reviewer 3: See step 1 for change recommendation

            Learn more about code scope when your scripts load JavaScript 
            into :binary:`~bin.mongosh`: :ref:`mongosh-scoping`.

            Or learn how to work with database query results in class 
            constructor functions, non-async generator functions, or callbacks

>             to ``.sort()`` on an array: :ref:`mongosh-script-considerations`.

Reviewer: I’m not sure if this is really something we want to talk about in a landing page? This is intended for users who may have existing scripts for the legacy shell that don’t work anymore in mongosh and doesn’t really target new users (which I assume the landing page here does).

Reviewer 3: +1 to leaving this out!

:hidefeedback: header
:noprevnext:

.. program:: mongosh

.. binary:: mongosh

.. _mdb-shell-overview:

======================================
Welcome to MongoDB Shell (``mongosh``)
======================================

.. introduction::

   The |mdb-shell|, ``mongosh``, is a JavaScript and Node.js |repl| 
   environment for interacting with MongoDB deployments in 
   `Atlas <https://www.mongodb.com/docs/atlas/>`__, locally,
   or on another remote host. Use the |mdb-shell| to test queries and 
   interact with the data in your MongoDB database.

   .. button:: Download mongosh
      :uri: https://www.mongodb.com/try/download/shell?jmp=docs

.. image:: /images/mongosh-plp/Technical_SOFTWARE_Terminal(3)_Spot.png
   :alt: Abstract illustration showing a terminal interacting with different systems 

.. kicker:: What You Can Do

Access MongoDB From Your Shell
------------------------------

.. tabs::

   .. tab:: Connect to MongoDB
      :tabid: connect

      .. procedure::

         .. step:: Find Your Connection String

            Find your connection string. The connection string varies 
            depending on the type of deployment you're connecting to.

>             Learn how to find your connection string on

Reviewer 2: Nit: change "on" to "for"
```suggestion
            Learn how to find your connection string for
```

            `Atlas <https://www.mongodb.com/docs/manual/reference/connection-string/#find-your-connection-string>`__.

            Or connect to an 

>             `on-premise <https://www.mongodb.com/docs/manual/reference/connection-string/#find-your-self-hosted-deployment-s-connection-string>`__

Reviewer 2: Changing "on-premise" to "self-hosted" per guidance I received from Server PMs.
```suggestion
            `self-hosted <https://www.mongodb.com/docs/manual/reference/connection-string/#find-your-self-hosted-deployment-s-connection-string>`__
```

            deployment.

         .. step:: Connect to MongoDB

>             Connect to a MongoDB deployment in Atlas using the Atlas 
>             connection string:

Reviewer 2: Suggestion: tweak language so that it's relevant to both Atlas and on prem, but keep the Atlas example
```suggestion
            Connect to a MongoDB deployment using the
            connection string. 
            
            The following connection string connects to 
            an Atlas deployment:
```

            explain queries with :ref:`Collection Methods 
            <mongosh-collection-methods>`.

         .. step:: Administer Servers

            Manage :ref:`replication <mongosh-replication-methods>` or 
            :ref:`sharding <mongosh-sharding-methods>` conveniently in 
            your shell.

            Check server status with a variety of :ref:`Server Status Methods 
            <mongosh-server-status-methods>`.

         .. step:: Manage Users and Roles

            Create or update roles, define and update privileges, or drop 
            roles using :ref:`Role Management Methods 
            <mongosh-role-management-methods>`.

            Create and update users, authenticate users, and manage user
            roles with :ref:`User Management Methods 
            <mongosh-user-management-methods>`. 

      .. image:: /images/mongosh-plp/Technical_SOFTWARE_Terminal(1)_Spot.png
         :alt: Abstract illustration showing a terminal performing operations

   .. tab:: Automate with Scripts
      :tabid: scripts

      .. procedure::

         .. step:: Run Scripts for CRUD or Administrative Tasks

>             Write scripts for the |mdb-shell| that perform CRUD or 

Reviewer 3: Reworded a bit for clarity:
"Write scripts to run with the MongoDB Shell that perform CRUD or administrative operations in MongoDB"

            administrative operations in MongoDB.

            For example, if you have a JS file that seeds synthetic or
            mock data into MongoDB in your development or staging 
            environment, run the file with:

            .. code-block:: sh

               mongosh YOUR_JS_FILENAME.js

            Explore a tutorial that uses the |mdb-shell| with JavaScript to
            access MongoDB: :ref:`mdb-shell-write-scripts`.

         .. step:: Create Custom Helper Functions with .mongoshrc

>             If you often find yourself converting date strings to ISO format

Reviewer 3: Suggestion:

Repeatedly writing large helper functions in the Shell? Store them in a [.mongoshrc config file](https://preview-mongodbdacharyc.gatsbyjs.io/mongodb-shell/DOCSP-32028/mongoshrc/#std-label-mongoshrc-js). For example, if you often find yourself converting date strings to ISO format for queries, create a function in .mongoshrc to handle it: ...

            environment, run the file with:

            .. code-block:: sh

               mongosh YOUR_JS_FILENAME.js

            Explore a tutorial that uses the |mdb-shell| with JavaScript to
            access MongoDB: :ref:`mdb-shell-write-scripts`.

         .. step:: Create Custom Helper Functions with .mongoshrc

            Repeatedly writing large helper functions in the Shell? Store 
            them in a :ref:`.mongoshrc config file <mongoshrc-js>`. For 
            example, if you often find yourself converting date strings to 
            ISO format for queries, create a function in ``.mongoshrc`` to 
            handle it:

            .. code-block:: JavaScript

               function toISO(dateString) { 
                  return new Date(dateString).toISOString(); 
               }

            Then, call the function in ``mongosh``:

            .. code-block:: sh

               db.clientConnections.find( { connectTime: toISO("06/07/2017") } )

            For more information, refer to :ref:`mongosh-write-scripts-config-file`.

         .. step:: Use or Publish Snippets

            :ref:`Pull existing snippets <snip-install-packages>` into 

>             your codebase for convenient reuse. Or :ref:`create and share 

Author: Applying readability heuristics, with these two sentences separated, they're each "grade 8" and qualify as "easy to read." If I combine them into a compound sentence, it becomes "grade 12" and "hard to read." In the docs sets I regularly work on, we would prioritize the readability and leave them separate. But if you think the compound sentence is more consistent with your style and tone, I could make this change. I'm happy to apply whatever would be most conventional for your team.

>             your codebase for convenient reuse. Or :ref:`create and share 

Reviewer 4: Ah gotcha, that makes sense! Definitely agree we should prioritize readability :)

            snippets <snip-create>` for your custom use case.

            For example, you might have a snippet that validates the data
            you import daily as a cron job. You can publish this snippet 
            so your development team can access it. Publish to a community
            registry, or configure a private registry. 
            
            For more information, refer to :ref:`snip-reg-config`.

      .. image:: /images/mongosh-plp/Technical_SOFTWARE_Terminal(4)_Spot.png
         :alt: Abstract illustration implying a terminal executing scripts

.. kicker:: Learn More

Other Powerful Features
-----------------------

.. card-group::
   :columns: 3
   :style: compact

   .. card::

>       :cta: Unlock the power of Editor Mode

Reviewer 4: [nit] For consistency with the other CTAs, should this be in title case?

      :url: https://www.mongodb.com/docs/mongodb-shell/reference/editor-mode/
      :icon: /images/mongosh-plp/icons/Technical_REALM_CLI10x.png
      :icon-alt: CLI Icon

>       Use an external editor or built-in editor to work with multiline

Reviewer 4: [suggestion] For brevity
```suggestion
      Use an external or built-in editor to work with multiline
```
